From 620339cf7b08ad7bf97b18b7ae7cf68c29c7ee7d Mon Sep 17 00:00:00 2001
From: Devarsh Thakkar <devarsh.thakkar@xilinx.com>
Date: Mon, 26 Aug 2019 09:41:59 -0700
Subject: [PATCH 0665/1851] v4l: xilinx: dma: Early dequeue first buffer for
 low latency capture

commit 4663a93173225aa6dcd85f767c9cd8957e024b04 from
https://github.com/Xilinx/linux-xlnx.git

Allow early dequeue for first buffer on streamon when low latency is
enabled as we require to send it to downstream component (for e.g. encoder)
so that it can be initialized.

When downstream component is initialized, application will send s_ctrl
with XVIP_START_DMA to start the dmaengine and buffer will start
getting filled with actual data.

Signed-off-by: Devarsh Thakkar <devarsh.thakkar@xilinx.com>
Reviewed-by: Hyun Kwon <hyun.kwon@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 drivers/media/platform/xilinx/xilinx-dma.c | 11 ++++++++++-
 1 file changed, 10 insertions(+), 1 deletion(-)

diff --git a/drivers/media/platform/xilinx/xilinx-dma.c b/drivers/media/platform/xilinx/xilinx-dma.c
index 2c1fcea932b8..5882358aa53f 100644
--- a/drivers/media/platform/xilinx/xilinx-dma.c
+++ b/drivers/media/platform/xilinx/xilinx-dma.c
@@ -686,8 +686,17 @@ static int xvip_dma_start_streaming(struct vb2_queue *vq, unsigned int count)
 	 * We dont't want to start DMA in case of low latency capture mode,
 	 * applications will start DMA using S_CTRL at later point of time.
 	 */
-	if (!dma->low_latency_cap)
+	if (!dma->low_latency_cap) {
 		dma_async_issue_pending(dma->dma);
+	} else {
+		/* For low latency capture, return the first buffer early
+		 * so that consumer can initialize until we start DMA.
+		 */
+		buf = list_first_entry(&dma->queued_bufs,
+				       struct xvip_dma_buffer, queue);
+		xvip_dma_complete(buf);
+		buf->desc->callback = NULL;
+	}
 
 	/* Start the pipeline. */
 	ret = xvip_pipeline_set_stream(pipe, true);
-- 
2.31.1

